Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderCrypto.cpp

```diff
@@ -0,0 +1,177 @@
+      0xDB5C, 0xEB3F, 0xFB1E, 0x8BF9, 0x9BD8, 0xABBB, 0xBB9A, 0x4A75, 0x5A54, 0x6A37, 0x7A16,
+      0x0AF1, 0x1AD0, 0x2AB3, 0x3A92, 0xFD2E, 0xED0F, 0xDD6C, 0xCD4D, 0xBDAA, 0xAD8B, 0x9DE8,
+      0x8DC9, 0x7C26, 0x6C07, 0x5C64, 0x4C45, 0x3CA2, 0x2C83, 0x1CE0, 0x0CC1, 0xEF1F, 0xFF3E,
+      0xCF5D, 0xDF7C, 0xAF9B, 0xBFBA, 0x8FD9, 0x9FF8, 0x6E17, 0x7E36, 0x4E55, 0x5E74, 0x2E93,
+      0x3EB2, 0x0ED1, 0x1EF0};
+
+  u16 crc = init_value;
+
+  for (size_t i = 0; i < data.size(); ++i)
+  {
+    const u16 tmp = (crc >> 8) ^ data[i];
+    crc = (crc << 8) ^ CRC_CCITT_TABLE[tmp];
+  }
+
+  return crc;
+}
+// CRC-64 algorithm that is limited to 48 bits every iteration
+u64 ComputeCRC48(std::span<const u8> data)
+{
+  const u64 polynomial = 0x42f0e1eba9ea3693;
+  const u64 initial_register_value = 2ULL * 2ULL * 3ULL * 1103ULL * 12868356821ULL;
+
+  u64 crc = initial_register_value;
+  for (size_t i = 0; i < data.size(); ++i)
+  {
+    crc ^= (static_cast<u64>(data[i]) << 40);
+    for (size_t j = 0; j < 8; ++j)
+    {
+      if (Common::ExtractBit(crc, 47))
+      {
+        crc = (crc << 1) ^ polynomial;
+      }
+      else
+      {
+        crc <<= 1;
+      }
+    }
+  }
+  return crc & 0x0000FFFFFFFFFFFF;
+}
+u64 CalculateKeyA(u8 sector, std::span<const u8, 0x4> nuid)
+{
+  if (sector == 0)
+  {
+    return 73ULL * 2017ULL * 560381651ULL;
+  }
+
+  std::array<u8, 5> data = {nuid[0], nuid[1], nuid[2], nuid[3], sector};
+
+  u64 big_endian_crc = ComputeCRC48(data);
+  u64 little_endian_crc = Common::swap64(big_endian_crc) >> 16;
+
+  return little_endian_crc;
+}
+void ComputeChecksum(ChecksumType type, const u8* data_start, u8* output)
+{
+  u16 crc = 0xFFFF;
+
+  switch (type)
+  {
+  case ChecksumType::Type0:
+  {
+    std::array<u8, 0x1E> input = {};
+    memcpy(input.data(), data_start, 0x1E);
+
+    crc = ComputeCRC16(crc, input);
+    break;
+  }
+  case ChecksumType::Type3:
+  {
+    std::array<u8, 0x110> input = {};
+    memcpy(input.data(), data_start, 0x20);
+    memcpy(input.data() + 0x20, data_start + 0x30, 0x10);
+
+    crc = ComputeCRC16(crc, input);
+    break;
+  }
+  case ChecksumType::Type2:
+  {
+    std::array<u8, 0x30> input = {};
+    memcpy(input.data(), data_start, 0x20);
+    memcpy(input.data() + 0x20, data_start + 0x30, 0x10);
+
+    crc = ComputeCRC16(crc, input);
+    break;
+  }
+  case ChecksumType::Type1:
+  {
+    std::array<u8, 0x10> input = {};
+    memcpy(input.data(), data_start, 0x10);
+    input[0xE] = 0x05;
+    input[0xF] = 0x00;
+
+    crc = ComputeCRC16(crc, input);
+    break;
+  }
+  case ChecksumType::Type6:
+  {
+    std::array<u8, 0x40> input = {};
+    memcpy(input.data(), data_start, 0x20);
+
+    memcpy(input.data() + 0x20, data_start + 0x30, 0x20);
+
+    input[0x0] = 0x06;
+    input[0x1] = 0x01;
+
+    crc = ComputeCRC16(crc, input);
+    break;
+  }
+  default:
+    break;
+  }
+  memcpy(output, &crc, 2);
+}
+void ComputeToyCode(u64 code, u8* output)
```

Comment:
This function returning a maximum of 10 bytes combined with the definition of `std::array<u8, 11> toy_code` over in `SkylanderData` makes this seem like a nullterminated string, but the null is never written here. You may just want to pass a `std::array<u8, 11>* output` here and make sure all the unused bytes after the end of the code are null. Ignore this if this isn't actually supposed to be nullterminated -- but it sure looks like it should be to me.

Reply:
Just wrote out the intermediate variable. Everything should now be null-terminated by default.

Reply:
Mm no this still looks off to me. I think you also introduced an off-by-one error in that loop a the bottom. This is probably confusing because you're reusing the same array for multiple purposes. Might I suggest instead something like:

```
void ComputeToyCode(u64 code, std::array<u8, 11>* output)
{
  if (code == 0)
  {
    static constexpr std::array<u8, 11> invalid_code_result{
        static_cast<u8>('N'), static_cast<u8>('/'), static_cast<u8>('A')};
    *output = invalid_code_result;
    return;
  }

  std::array<u8, 10> code_bytes;
  for (size_t i = 0; i < code_bytes.size(); ++i)
  {
    code_bytes[i] = static_cast<u8>(code % 29);
    code /= 29;
  }

  static constexpr char lookup_table[] = "23456789BCDFGHJKLMNPQRSTVWXYZ";
  std::array<u8, 10> code_chars;
  for (size_t i = 0; i < code_bytes.size(); ++i)
  {
    code_chars[i] = static_cast<u8>(lookup_table[code_bytes[9 - i]]);
  }

  std::array<u8, 11> result;
  std::memcpy(&result[0], &code_chars[0], 5);
  result[5] = static_cast<u8>('-');
  std::memcpy(&result[6], &code_chars[5], 5);

  *output = result;
}
```

Reply:
Your implementation is definitely more readable. I also added a test to be able to spot mistakes like this easier